### PR TITLE
Add MIME type based file extensions to snapshot data files [SE-4663] 

### LINKED ISSUE #63: Blockstore can't serve SVG images correctly over S3
When one uploads an SVG image to a Blockstore server that's storing data on Amazon S3, Blockstore will serve the image from a URL like

```
https://our-blockstore-content.s3.amazonaws.com/ca9635e6-a43e-40df-be67-1764d4a6b311/snapshot_data/6bc0efe8f1d9722615ca1d7042b85d5af115ca6f?AWSAccessKeyId=...&Expires=1575574652
```

S3 in turn serves that response with `Content-Type: application/octet-stream`

Unfortunately, this means that one cannot use `` because browsers will not render SVG images that aren't served with the correct `Content-Type` header.

Possible solutions include:
* Change the data file name so it includes the file extension
* Have Blockstore know/detect the file type, and then [include the `response-content-type` parameter in the S3 signed URL request](https://stackoverflow.com/a/36306039), so S3 will serve it with the correct `Content-Type` header.

### PR DESCRIPTION
**Pull request by jvdm**
## Description

Change Draft and Version store metadata to track static file's MIME type based on its "file extension" as defined by [`mimetypes.guess_type()`](https://docs.python.org/3/library/mimetypes.html#mimetypes.guess_type). Additionally, sets the extension based on the guessed mime type when saving files in the Version Store, in an attempt to allow correct MIME type headers when serving the files.

Finally, the change is backward compatible. If a file's MIME type cannot be guessed from its extension or the file doesn't have an extension at all, it is stored using the old naming schema (that is, only the file's hash). This also allows files stored with the old metadata that doesn't carry the `mime_type` field to be handled correctly.

See additional discussion in the related issue #63.

## Author Comments, Concerns, and Open Questions

The PR approach fixes #63, but it doesn't attempt to conciliate old files with new files with the same hash content. That means, from the Blockstore perspective these are two different objects, although, in practice, they hold the same content.

## Test Instructions

1. Deploy the blockstore to a devstack as described in README.
2. Go to http://localhost:18250/api/v1
3. Use the APIs to (a.) Create a bundle, (b.) Create a draft, (c.) Then commit to the version. *Make sure to add static files*, for example, a *.jpg python.
4. Call `make blockstore-shell`.
5. Then check the following:
    - Go to `/blockstore/app/blockstore/media/<version>/snapshot_data` and verify files were added with their extension.
    - Read `/blockstore/app/blockstore/media/<version>/snapshots/<snapshot_id>.json` and verify the `content_type` flag is present.
    - Go to the files URL and verify the devstack serves the file with the right MIME type header.

## TODOs

- [ ] Squash before merging

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @jvdm! I've created [OSPR-5953](https://openedx.atlassian.net/browse/OSPR-5953) to keep track of it in JIRA, where we prioritize reviews. Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

**Comment by natabene** [comment]
@jvdm Thank you for your contribution. Please let me know once this is ready for a review.

**Comment by jvdm** [comment]
Thank you, @natabene. I believe this is ready for review. Although, some discussion may happen at #63 regarding the approach taken here.

**Comment by bradenmacdonald** [comment]
@ormsbee Would you like to review this? It is mostly touching on the code that you originally wrote I believe.

I am happy to review in addition or instead.

File: blockstore/apps/bundles/tests/test_store.py
Comment (arbrandes) [inline blockstore/apps/bundles/tests/test_store.py:122]: @jvdm, do you mind adding a file (or two) here that fall under the situation [Braden originally described](https://github.com/edx/blockstore/issues/63#issuecomment-893870623) where "two files are uploaded with different extensions but the same content"?  In particular, when such files resolve to the exact same MIME type.

If for nothing else other than to document that this was foreseen, and to protect this particular case from future bugs.  I couldn't detect a related problem just by reading the code.

**Review by arbrandes — CHANGES_REQUESTED** [review]
@jvdm, I generally approve of the PR.  I tested it in a devstack with real data, and it solves the problem as originally reported.  Congrats, and thanks!

I just asked for one change regarding tests.

**Comment by arbrandes** [comment]
Also, can you please add [SE-4663] to the title, to aid our internal tracking?

File: blockstore/apps/bundles/tests/test_store.py
Comment (jvdm) [reply blockstore/apps/bundles/tests/test_store.py:122]: @arbrandes 

> do you mind adding a file (or two) here that fall under the situation Braden originally described where "two files are uploaded with different extensions but the same content"?

Can do. Although, I want to clarify what will happen.

> In particular, when such files resolve to the exact same MIME type.

Two scenarios, (a.) If the different extensions are in fact the same MIME, the files will dedup in the storage layer, just like today. That is correct because the extensions are the source of truth for the MIME-type header in the patch proposed, so one potential example is `cat.jpeg` and `samecat.jpg` will be stored as `<some-hash>.jpg`. Now, (b.) if different extensions are guessed to be two different MIME types, for example, if you do `mv cat.jpg cat.pdf` and upload `cat.pdf` and `cat.jpeg`, blockstore will create two files with the same content.

You might argue that for (b.), we would be better using the files' content as the source of the truth for the MIME header. That is possible, but in my opinion, not as flexible as using the file extension. One example, what if the content-based MIME type guessing fails for some (I know, very particular) case? Customers will feel confused, as in "Why my PDF is opening as an image?". And without a clear path to move forward. So, unless we add some additional API parameters to bypass the guessing, I prefer the file extension. Let me know if you want to explore the API parameter to select the type forcibly.

Comment (bradenmacdonald) [reply blockstore/apps/bundles/tests/test_store.py:122]: For what it's worth, I think what you're describing in cases (a) and (b) sound good and I was thinking along the same lines too.

**Comment by bradenmacdonald** [comment]
@natabene Can I review this as core committer?

**Comment by bradenmacdonald** [comment]
@arbrandes Did you also test backwards compatibility ?

File: blockstore/apps/bundles/store.py
Comment (bradenmacdonald) [inline blockstore/apps/bundles/store.py:52]: I think I see the potential for a rare but possible problem here:

Scenario:
1. User uploads a video file; it's `mime_type` is stored as `video/mp4` and its data file is stored at `[hash].mp4`
2. We update python (or the underlying OS used in the docker image, etc.), and `mimetypes` now returns `.m4v` as the file extension for `video/mp4`
3. Blockstore hasn't stored the file extension, only the mimetype, so now looks for `[hash].mp4` which doesn't exist.


Obviously this isn't super likely, but it has happened before with mimetype projects, e.g. [this change of extension for `audio/mp4` from `.mp4a` to `.m4a`](https://github.com/jshttp/mime-db/commit/ae413e174e9b8d3026b39a43a9a044b72be41ab8) or [this very recent change of the extension for `application/ecmascript` from `.ecma` to `.es`](https://github.com/jshttp/mime-db/commit/03d57a35d71f5897f3a2a9df488c6ef76f34ccc6)

Another related issue is that `mimetypes` can return different results on different systems since it uses data files from the underlying OS.

All that said, it's probably not really a problem in practice and we could always add blockstore-specific overrides for certain mimetypes here if it ever was a problem. Thoughts?

**Comment by bradenmacdonald** [comment]
Code looks good to me, just the one question above. If @arbrandes has carefully tested backwards compatibility I'll approve this; otherwise I will as soon as he or I has done so.

**Comment by arbrandes** [comment]
@bradenmacdonald, I have just deployed this branch to a pre-existing sandbox that has Blockstore using MinIO and a large number of previously-imported blocks, and could detect no ill effects.  Previously uploaded SVGs still don't work, of course, but other assets such as images, and otherwise the bundle itself, are all fine.

**Review by bradenmacdonald — APPROVED** [review]

**Comment by bradenmacdonald** [comment]
Please squash to a single commit and we can merge next week if there are no further comments or concerns.

File: blockstore/apps/bundles/store.py
Comment (jvdm) [reply blockstore/apps/bundles/store.py:52]: @bradenmacdonald:

> I think I see the potential for a rare but possible problem here:
>
> [...]
>
> Blockstore hasn't stored the file extension, only the mimetype, so now looks for [hash].mv4 which doesn't exist.

Oh, that's a nice catch! Loved it. :heart: 

> Thoughts?

This dependency between blockstore behavior and its environment will be painful to maintain. A deployment or upgrade could easily break multiple static file URLs. For example, `XBlockRuntime` relies on [blockstore to translate static URLs](https://github.com/edx/edx-platform/blob/49aaa8621d195a452000942aae8286966a91b96a/openedx/core/djangoapps/xblock/runtime/blockstore_runtime.py#L168) which in turn will use the [snapshot URL generation](https://github.com/edx/blockstore/blob/ece89fa79cd4e4a320f2d8c977032af0a52e2680/blockstore/apps/api/v1/serializers/bundles.py#L152) to retrieve it, and if the extension guessed changes, let's say, because of deployment, so does the static file URL, yielding 400's.

It should be relatively simple to extend the patch to add the "guessed file extension" to the snapshot metadata. But, at this point, I think the patch will be more resilient if we go all the way and save the whole "snapshot filename" instead. Changes to the naming schema won't affect previous Snapshots if we rely on a static filename definition in the snapshot metadata.

Let's get a consensus on this before I submit a new revision, @bradenmacdonald.

**Comment by ormsbee** [comment]
> @ormsbee Would you like to review this? It is mostly touching on the code that you originally wrote I believe.

Sorry for the delay in reply. I'm on PTO for the next week. I trust you folks. Thank you.

**Comment by jvdm** [comment]
> Please squash to a single commit and we can merge next week if there are no further comments or concerns.

Rebased onto `origin/master` and squashed.

File: blockstore/apps/bundles/store.py
Comment (bradenmacdonald) [reply blockstore/apps/bundles/store.py:52]: @jvdm Whoops sorry for the delayed reply here. 

Do you want to merge this now or address this first? I'm fine either way. Although it does seem like a risk of a bug in the future, the odds strike me as very low, and the mitigation would be quite easy (`if mimetype='problematic/type': file_ext = ".forced"; else: file_ext = guess_mime_type()`)

>  I think the patch will be more resilient if we go all the way and save the whole "snapshot filename" instead

Sounds good to me

Comment (jvdm) [reply blockstore/apps/bundles/store.py:52]: @bradenmacdonald:

> Do you want to merge this now or address this first? I'm fine either way.

Let's merge. We can unblock customers now and follow up on this in a separate PR.

Comment (bradenmacdonald) [reply blockstore/apps/bundles/store.py:52]: OK I'll merge tomorrow so that it deploys during edX business hours.

**Comment by edx-webhook** [comment]
@jvdm 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/blockstore&entry.867055334=https://github.com/edx/blockstore/pull/118&entry.1484655318=https://github.com/jvdm&entry.752974735=2021-08-06+09:14&entry.1917517419=2021-08-19+15:24&entry.2133058324=Yes) so we can improve your experience in the future.

**Pull request merged**
